src/ucs/datastruct/ptr_map.inl

```diff
@@ -0,0 +1,125 @@
+/**
+* Copyright (C) Mellanox Technologies Ltd. 2020.  ALL RIGHTS RESERVED.
+*
+* See file LICENSE for terms.
+*/
+
+#ifndef UCS_PTR_MAP_INL_
+#define UCS_PTR_MAP_INL_
+
+#include "ptr_map.h"
+
+BEGIN_C_DECLS
+
+KHASH_IMPL(ucs_ptr_map_impl, ucs_ptr_map_key_t, void*, 1,
+           kh_int64_hash_func, kh_int64_hash_equal);
+
+/**
+ * Initialize a pointer map
+ *
+ * @param [in]  map     Map to initialize.
+ * @return      UCS_OK on success otherwise error code as defined by
+ *              @ref ucs_status_t.
+ */
+static inline ucs_status_t ucs_ptr_map_init(ucs_ptr_map_t *map)
+{
+    map->next_id = 0;
+    kh_init_inplace(ucs_ptr_map_impl, &map->hash);
+    return UCS_OK;
+}
+
+/**
+ * Destroy a pointer map
+ *
+ * @param [in]  map     Map to destroy.
+ */
+static inline void ucs_ptr_map_destroy(ucs_ptr_map_t *map)
+{
+    kh_destroy_inplace(ucs_ptr_map_impl, &map->hash);
+}
+
+/**
+ * Put a pointer into the map
+ *
+ * @param [in]  map     Container.
+ * @param [in]  ptr     Object pointer.
+ * @param [in]  direct  if not 0, @a key cannot be directly converted into
+ *                      @a ptr. Suggested to enable for extra @a ptr guard but
+ *                      it can impact performance.
+ * @param [out] key     Key to object pointer @ptr if operation completed
+ *                      successfully otherwise value is undefined.
+ * @note @ptr must be aligned on @ref UCS_PTR_MAP_KEY_MIN_ALIGN.
+ */
+static UCS_F_ALWAYS_INLINE ucs_status_t
+ucs_ptr_map_put(ucs_ptr_map_t *map, void *ptr, int indirect,
+                ucs_ptr_map_key_t* key)
+{
+    khiter_t iter;
+    int ret;
+
+    if (ucs_likely(!indirect)) {
+        *key = (uintptr_t)ptr;
+        ucs_assert(!(*key & UCS_PTR_MAP_KEY_MIN_ALIGN));
+        return UCS_OK;
+    }
+
+    *key = (map->next_id += UCS_PTR_MAP_KEY_MIN_ALIGN) |
+           UCS_PTR_MAP_KEY_INDIRECT_FLAG;
+
+    iter = kh_put(ucs_ptr_map_impl, &map->hash, *key, &ret);
+    if (ucs_unlikely(ret == UCS_KH_PUT_FAILED)) {
+        return UCS_ERR_NO_MEMORY;
+    }
+
+    if (ucs_unlikely(ret == UCS_KH_PUT_KEY_PRESENT)) {
+        return UCS_ERR_ALREADY_EXISTS;
```

Comment:
didn't we decide to have ucs_bug/ucs_fatal/assert in this case?
for example, ucs_ptr_map_del has ucs_assert in case value is not found